### PR TITLE
Add config transport endpoint

### PR DESCRIPTION
**Pull request by lloeki**
**What does this PR do?**

Add `/v0-7/config` endpoint and corresponding transport.

**Motivation**

Remote configuration.

**Additional Notes**

A dedicated transport is required, see #2676 

**How to test the change?**

The code is isolated and uncalled, so CI should stay green.

Manual testing:

- Step 1: start the agent
```
docker run --rm -it -e DD_APM_ENABLED=true -e DD_ENV=${YOUR_ENV} -e DD_HOSTNAME=${YOUR_HOSTNAME} -e DD_BIND_HOST=0.0.0.0 -e DD_APM_NON_LOCAL_TRAFFIC=true -e DD_API_KEY=${YOUR_API_KEY}-e DD_REMOTE_CONFIGURATION_ENABLED=true -p 8126:8126 datadog/agent:7.42.0
```

- Step 2: Make sure your API has the RC enable on the Org settings page
- Step 3: hit it with the following code

```
require 'ddtrace'

Datadog.configure do |c|
    c.agent.host = "#{YOUR_AGENT_HOST}   
end

transport_options = {
  agent_settings: Datadog::Core::Configuration::AgentSettingsResolver.call(Datadog.configuration)
}

require 'datadog/core/transport/http'

transport_v7 = Datadog::Core::Transport::HTTP.v7(**transport_options.dup)

CAP_ASM_ACTIVATION                = 1 << 1 # Remote activation via ASM_FEATURES product
CAP_ASM_IP_BLOCKING               = 1 << 2 # accept IP blocking data from ASM_DATA product
CAP_ASM_DD_RULES                  = 1 << 3 # read ASM rules from ASM_DD product
CAP_ASM_EXCLUSIONS                = 1 << 4 # exclusion filters (passlist) via ASM product
CAP_ASM_REQUEST_BLOCKING          = 1 << 5 # can block on request info
CAP_ASM_RESPONSE_BLOCKING         = 1 << 6 # can block on response info
CAP_ASM_USER_BLOCKING             = 1 << 7 # accept user blocking data from ASM_DATA product
CAP_ASM_CUSTOM_RULES              = 1 << 8 # accept custom rules
CAP_ASM_CUSTOM_BLOCKING_RESPONSE  = 1 << 9 # supports custom http code or redirect sa blocking response

capabilities = [
  CAP_ASM_IP_BLOCKING,
  CAP_ASM_USER_BLOCKING,
  CAP_ASM_CUSTOM_RULES,
  CAP_ASM_EXCLUSIONS,
  CAP_ASM_REQUEST_BLOCKING,
  CAP_ASM_RESPONSE_BLOCKING,
  CAP_ASM_DD_RULES,
].reduce(&:|)

capabilities_binary = capabilities
  .to_s(16)
  .tap { |s| s.size.odd? && s.prepend('0') }
  .scan(/\h\h/)
  .map { |e| e.to_i(16) }
  .pack('C*')

products = [
  'ASM_DD',       # Datadog employee issued configuration
  'ASM',          # customer issued configuration (rulesets, passlist...)
  'ASM_FEATURES', # capabilities
  'ASM_DATA',     # config files (IP addresses or users for blocking)
]

state = OpenStruct.new(
  {
    root_version: 1,              # unverified mode, so 1
    targets_version: 0,           # from scratch, so zero
    config_states: [],            # from scratch, so empty
    has_error: false,             # from scratch, so false
    error: '',                    # from scratch, so blank
    opaque_backend_state: '',     # from scratch, so blank
  }
)

id = SecureRandom.uuid # client id

payload = {
  client: {
    state: {
      root_version:         state.root_version,
      targets_version:      state.targets_version,
      config_states:        state.config_states,
      has_error:            state.has_error,
      error:                state.error,
      backend_client_state: state.opaque_backend_state,
    },
    id: id,
    products: products,
    is_tracer: true,
    is_agent: false,
    client_tracer: {
      runtime_id:      Datadog::Core::Environment::Identity.id,
      language:        Datadog::Core::Environment::Identity.lang,
      tracer_version:  Datadog::Core::Environment::Identity.tracer_version,
      service:         Datadog.configuration.service,
      env:             Datadog.configuration.env,
    # app_version:     app_version,   # TODO: I don't know where this is in the tracer
      tags:            [],            # TODO: add nice tags!
    },
    # base64 is needed otherwise the Go agent fails with an unmarshal error
    capabilities: Base64.encode64(capabilities_binary).chomp,
  },
  cached_target_files: [
  # {
  #   path: '',
  #   length: 0,
  #   hashes: '';
  # }
  ],
}


res = transport_v7.send_config(payload)
puts res.target_files
puts res.client_configs
puts res.targets
puts res.roots
```

### CODE REVIEW

**Comment by lloeki** [comment]
Pending:

- specs
- RBS typing

File: lib/datadog/core/transport/config.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+          ##### there is only one transport! it's negotiation!
+          def send_config(payload)
+            json = JSON.dump(payload)
+            parcel = EncodedParcel.new(json)
+            request = Request.new(parcel)
+
+            response = @client.send_config_payload(request)
+
+            # TODO: not sure if we're supposed to do that as we don't chunk like traces
+            # Datadog.health_metrics.transport_chunked(responses.size)
```
Comment (marcotc) [inline lib/datadog/core/transport/config.rb:53]: Feel free to skip this then, no worries.

File: lib/datadog/core/transport/http/config.rb [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+              # TODO: these fallbacks should be improved
+              roots = options[:roots] || []
+              targets = options[:targets] || ''
+              target_files = options[:target_files] || []
+              client_configs = options[:client_configs] || []
+
+              raise TypeError.new(Array, roots) unless roots.is_a?(Array)
+
+              @roots = roots.map do |e|
```
Comment (GustavoCaso) [inline lib/datadog/core/transport/http/config.rb:49]:
```suggestion
              @roots = roots.map do |root|
```

File: lib/datadog/core/transport/http/config.rb [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+                parsed
+              end
+
+              raise TypeError.new(String, targets) unless targets.is_a?(String)
+
+              @targets = begin
+                begin
+                  decoded = Base64.strict_decode64(targets)
+                rescue ArgumentError
```
Comment (GustavoCaso) [inline lib/datadog/core/transport/http/config.rb:74]: `rescue ArgumentError => e`

File: lib/datadog/core/transport/http/config.rb [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+                  raise DecodeError.new(:targets, e)
+                end
+
+                begin
+                  parsed = JSON.parse(decoded) # TODO: unprocessed, don't symbolize_names
+                rescue JSON::ParserError
```
Comment (GustavoCaso) [inline lib/datadog/core/transport/http/config.rb:80]: `rescue JSON::ParserError => e`

File: lib/datadog/core/transport/http/config.rb (hunk 2 above)
Comment (GustavoCaso) [reply lib/datadog/core/transport/http/config.rb:49]: And I would change all the subsequent references of `e` to `root`

File: lib/datadog/core/transport/http/config.rb [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+                  decoded = Base64.strict_decode64(e) # TODO: unprocessed, don't symbolize_names
+                rescue ArgumentError
+                  raise DecodeError.new(:roots, e)
+                end
+
+                begin
+                  parsed = JSON.parse(decoded)
+                rescue JSON::ParserError
+                  raise ParseError.new(:roots, e)
+                end
```
Comment (GustavoCaso) [inline lib/datadog/core/transport/http/config.rb:62]: Do we want to capture the specific error here, or do we want to provide the `root` instance?

File: lib/datadog/core/transport/http/config.rb [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+                begin
+                  parsed = JSON.parse(decoded) # TODO: unprocessed, don't symbolize_names
+                rescue JSON::ParserError
+                  raise ParseError.new(:targets, e)
+                end
+
+                # TODO: perform more processing to validate content. til then, no freeze
+
+                parsed
+              end
```
Comment (GustavoCaso) [inline lib/datadog/core/transport/http/config.rb:87]:
```suggestion
 raise TypeError.new(String, targets) unless targets.is_a?(String)

begin
  decoded_targets = Base64.strict_decode64(targets)
rescue ArgumentError => e
  raise DecodeError.new(:targets, e)
end

begin
  parsed_targets = JSON.parse(decoded_targets) # TODO: unprocessed, don't symbolize_names
rescue JSON::ParserError => e
  raise ParseError.new(:targets, e)
end

# TODO: perform more processing to validate content. til then, no freeze
@targets = parsed_targets
```

**Comment by GustavoCaso** [comment]
We can probably extract some methods for parsing `roots`, `targets` and `target_files`. Also, error handling could improve we could create helper methods for `json` parsing and `decode`

File: lib/datadog/core/transport/http/config.rb (hunk 5 above)
Comment (lloeki) [reply lib/datadog/core/transport/http/config.rb]: the root value (see the `ParseError#initialize` arguments)

File: lib/datadog/core/transport/http/config.rb (hunk 6 above)
Comment (lloeki) [reply lib/datadog/core/transport/http/config.rb]: Not sure what this suggestion is about, care to elaborate?

Comment (GustavoCaso) [reply lib/datadog/core/transport/http/config.rb]: Removing the nested `begin` blocks

**Review by GustavoCaso — APPROVED** [review]

**Pull request merged**